### PR TITLE
Add tidal astronomy, equilibrium forcing, and boundary conditions

### PR DESCRIPTION
**Pull request by enri66**
## Summary

Adds the generic machinery to force and open-boundary-drive the astronomical tide, independent of any particular tidal atlas:

- `TidalHarmonics(reference_date; constituents, ramp_time)`: frequencies, phases, and nodal factors of the ten primary tidal constituents (`M2`, `S2`, `N2`, `K2`, `K1`, `O1`, `P1`, `Q1`, `Mf`, `Mm`), from Schureman (1958) / Kowalik and Luick (2019).
- `tidal_forcing(harmonics)`: a barotropic `(u, v)` forcing from the equilibrium tide's pressure gradient.
- `tidal_boundary_conditions(grid, harmonics, dataset; kw...)`: Flather (1976) conditions on `U`/`V` and Chapman (1985) radiation on `η`, sampling `dataset` through the new `tidal_atlas_constants` extension point (implemented for a specific atlas downstream, e.g. TPXO, in NumericalEarth.jl).

A single `TidalHarmonics` instance feeds both the body force and the boundary conditions so they stay in phase.

## Test plan

- [x] New unit tests, `Tidal astronomy` (frequencies/phases checked against NOAA CO-OPS published angular speeds) and `Equilibrium tidal body force` (a small basin's modeled response compared against the analytical equilibrium tide), added to test_open_boundary_conditions_hydrostatic.jl.
- [x] Full Open Boundary Conditions for HydrostaticFreeSurfaceModel testset passes, 11/11.
- [x] Downstream integration verified against a synthetic tidal atlas and against NOAA tide gauges in a regional Mid-Atlantic Bight configuration (NumericalEarth.jl).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by simone-silvestri** [comment]
I think you want to also modify the split explicit scheme here. We pass a fixed clock to the split explicit routine, but since tides are time dependent we might want to update the clock also inside the substepping routine so that the OBCs on the barotropic variables are computed on the correct time

**Comment by enri66** [comment]
You're right — traced it through `iterate_split_explicit!`: `substep_clock` is built once before the substep loop with `time = clock.time` (the macro step's start time), then baked into the pre-converted halo-fill argument tuples and reused unchanged across all `Nsubsteps`. Since `tidal_transport_and_elevation` reads `clock.time` fresh on every call, the tidal phase gets evaluated once per substep but always at the same stale time.

This looks general to split-explicit substepping rather than specific to tides — any time-dependent barotropic OBC would hit it. The substep clock is hoisted and GPU-converted outside the loop on purpose (avoiding per-substep conversion overhead across ~50 tiny kernels), so a fix that doesn't reintroduce that cost means threading the scalar substep time the same lightweight way `averaging_weight`/`transport_weight` already are, rather than just bumping `substep_clock.time` in place — that touches the shared substep plumbing more than this PR's files do.

Would you be OK with a follow-up issue/PR for that instead of folding it into this one? Happy to open it.

**Comment by enri66** [comment]
Follow-up PR: https://github.com/CliMA/Oceananigans.jl/pull/5982

File: src/BoundaryConditions/open_boundary_schemes/tidal_forcing.jl [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+    S2 = (amplitude = 0.112743 * 0.693, argument = ( 2,  0,  0,  0,  0), nodal = (f₀ = 1.000, f₁ =  0.000, u₁ =   0.0)),
+    N2 = (amplitude = 0.046397 * 0.693, argument = ( 2, -3,  2,  1,  0), nodal = (f₀ = 1.000, f₁ = -0.037, u₁ =  -2.1)),
+    K2 = (amplitude = 0.030684 * 0.693, argument = ( 2,  0,  2,  0,  0), nodal = (f₀ = 1.024, f₁ =  0.286, u₁ = -17.7)),
+    K1 = (amplitude = 0.141565 * 0.736, argument = ( 1,  0,  1,  0,  1), nodal = (f₀ = 1.006, f₁ =  0.115, u₁ =  -8.9)),
+    O1 = (amplitude = 0.100661 * 0.695, argument = ( 1, -2,  1,  0, -1), nodal = (f₀ = 1.009, f₁ =  0.187, u₁ =  10.8)),
+    P1 = (amplitude = 0.046848 * 0.706, argument = ( 1,  0, -1,  0, -1), nodal = (f₀ = 1.000, f₁ =  0.000, u₁ =   0.0)),
+    Q1 = (amplitude = 0.019273 * 0.695, argument = ( 1, -3,  1,  1, -1), nodal = (f₀ = 1.009, f₁ =  0.187, u₁ =  10.8)),
+    Mf = (amplitude = 0.042041 * 0.693, argument = ( 0,  2,  0,  0,  0), nodal = (f₀ = 1.043, f₁ =  0.414, u₁ = -23.7)),
+    Mm = (amplitude = 0.022191 * 0.693, argument = ( 0,  1,  0, -1,  0), nodal = (f₀ = 1.000, f₁ = -0.130, u₁ =   0.0)),
```
Comment (glwagner) [inline src/BoundaryConditions/open_boundary_schemes/tidal_forcing.jl]: this smells like NumericalEarth stuff (it is specific to Earth for example!) although I can be convinced that it doesn't belong there for some reason. Does it have to integrate deeply into the OBCs?

File: src/BoundaryConditions/open_boundary_schemes/tidal_forcing.jl [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+Cartwright–Tayler amplitude times the Love number factor ``1 + k - h`` of an elastic Earth; the
+multiples of the angles in `astronomical_angles` that sum to the equilibrium argument ``V``; and the
+`nodal` coefficients, with which the 18.6-year cycle of the lunar node longitude ``N`` modulates
+amplitude as ``f = f₀ + f₁ \\cos N`` and phase as ``u = u₁ \\sin N`` [°].
+
+The first multiple is the constituent's species: 2 semidiurnal, 1 diurnal, 0 long-period.
+
+References: Schureman (1958), [Kowalik and Luick (2019)](@cite kowalik2019modern).
+"""
+const tidal_constituents = (
```
Comment (glwagner) [inline src/BoundaryConditions/open_boundary_schemes/tidal_forcing.jl]: note these are specific to Earth, so might be good to indicate so --- `earth_tidal_constituents`. In fact, they are also specific to the present day! Also rather than hard-coding this as a const, it may be better to somehow pass a data structure in as an argument. That will allow it to be changed for other celestial situations.

File: src/BoundaryConditions/open_boundary_schemes/tidal_forcing.jl [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+using Dates
+
+TidalHarmonics(DateTime(2019, 4, 1); constituents = (:M2, :K1))
+
+# output
+TidalHarmonics at 2019-04-01T00:00:00 with M2, K1
+```
+"""
+function TidalHarmonics(reference_date::DateTime;
+                        constituents = keys(tidal_constituents),
```
Comment (glwagner) [inline src/BoundaryConditions/open_boundary_schemes/tidal_forcing.jl]: so the `constituents` is a tuple of symbols? One concern I have is that the symbols are a bespoke notation for Earth so this is somehow not a general design (unfortunately, not a very helpful comment since I am not sure what a general/planet-agnostic design would look like). It might be nice to search for it though.

File: src/BoundaryConditions/open_boundary_schemes/tidal_forcing.jl [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+                        ramp_time = 0)
+
+    names = Tuple(Symbol(constituent) for constituent in constituents)
+    properties = Tuple(tidal_constituents[name] for name in names)
+
+    centuries = Dates.value(reference_date - astronomical_epoch) / (86_400_000 * 36_525)
```
Comment (glwagner) [inline src/BoundaryConditions/open_boundary_schemes/tidal_forcing.jl]: here again I think are numbers specific to Earth?

Comment (enri66) [reply src/BoundaryConditions/open_boundary_schemes/tidal_forcing.jl]: @glwagner let me see if I can generalize it.  I was unsure where this belongs, and my first intuition was to put it in NumericalEarth, then moved it here after some discussion.  I see the point though.

Comment (enri66) [reply src/BoundaryConditions/open_boundary_schemes/tidal_forcing.jl]: @glwagner ...I think it's doable:   The latitude-based species structure (cos²φ, sin2φ, etc.) isn't actually Earth-specific — it's just spherical-harmonic geometry for any tide-raising body on a rotating sphere. So the genuinely general parts are the harmonic summation machinery, nodal modulation, ramping, equilibrium-elevation gradient, boundary condition plumbing, and the atlas-sampling interface — all of which could work for other celestial bodies given different physical constants.  So, let me see.

**Comment by enri66** [comment]
@glwagner you were right, and generalizing it turned out to be mostly deletion. Pushed in f924c0bf3.

It helped to notice that `TidalHarmonics` was *already* the general object: it stores frequencies, phases, nodal factors, equilibrium amplitudes and species, and every consumer downstream — the equilibrium tide, the body force, the Flather condition — reads only those. All the Earth was concentrated in the constructor that derived them from a date. So the fix was to split the constructor, not to redesign anything:

```julia
TidalHarmonics(; constituents, frequencies, phases, equilibrium_amplitudes, species,
                 nodal_factors = ..., ramp_time = 0)
```

Gone from Oceananigans: the constituent table, the 1900 epoch, the mean longitudes and their per-Julian-century rates, the lunar-node modulation, and the `36_525 * 86_400` that baked in Earth's day. Frequencies now arrive in rad/s, so no epoch or century unit survives here. Net −66 lines.

Two things I kept, and why:

- **`species`.** It sets the latitude structure — cos²φ, sin2φ, ½−3⁄2sin²φ — which is the degree-2 decomposition of the potential of any distant body, not an Earth table. Mars gets the same three.
- **`constituents`.** Now just opaque labels passed through to `tidal_atlas_constants`, i.e. whatever your atlas calls its constituents. Oceananigans attaches no meaning to them, which I think answers your "bespoke notation" concern — they're no longer an index into an Earth table.

The Earth side moves to NumericalEarth as `earth_tidal_harmonics(reference_date; constituents, ramp_time)`, next to `TPXO10Atlas`, which is the dataset that pairs with it ([NumericalEarth#681](https://github.com/NumericalEarth/NumericalEarth.jl/pull/681)). I made it a named function rather than a `TidalHarmonics(::DateTime)` method since NumericalEarth owns neither the type nor `DateTime`. It reproduces all ten NOAA CO-OPS published angular speeds to `rtol = 1e-7`, and that check moved over with it — so the astronomy is still tested, just where the astronomy now lives.

On your first question — whether it has to integrate deeply into the OBCs — what's left here is the Flather/Chapman plumbing plus `tidal_atlas_constants`, which is the hook an atlas implements. That part is dataset- and planet-agnostic now. Happy to move more of it if you still think it belongs elsewhere.

**Comment by enri66** [comment]
@glwagner , one more note, PR #681 in NumericalEarth won't run until this one merges...a consequence of the split, but I think it makes sense.  Europa here we come!

File: src/BoundaryConditions/open_boundary_schemes/tidal_forcing.jl [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+@inline tidal_ramp(harmonics, t) = ifelse(harmonics.ramp_time > 0, tanh(t / harmonics.ramp_time), one(t))
+
+#####
+##### The equilibrium tide
+#####
+
+# Elevation [m] of the equilibrium tide at a cell center. The latitude structure of a constituent
+# follows from its species: cos²φ semidiurnal, sin2φ diurnal, and ½ - 3/2 sin²φ long-period.
+@inline function equilibrium_tide(i, j, k, grid, clock, harmonics)
```
Comment (simone-silvestri) [inline src/BoundaryConditions/open_boundary_schemes/tidal_forcing.jl:90]: is this function expensive?

File: test/hydrostatic_free_surface/open_boundary_conditions.jl [hunk 6]
```diff
@@ -546,6 +547,67 @@ function test_oblique_radiation_mirror_symmetry()
...
+    period = 2π / ω
+
+    harmonics = TidalHarmonics(constituents = (:M2,),
+                               frequencies = (ω,),
+                               phases = (1.7,),
+                               equilibrium_amplitudes = (0.168,),
+                               species = (2,),
+                               ramp_time = period)
+
+    model = HydrostaticFreeSurfaceModel(grid;
```
Comment (simone-silvestri) [inline test/hydrostatic_free_surface/open_boundary_conditions.jl:577]: would we need boundary conditions here?

File: src/BoundaryConditions/open_boundary_schemes/tidal_forcing.jl (hunk 5 above)
Comment (enri66) [reply src/BoundaryConditions/open_boundary_schemes/tidal_forcing.jl:90]: I can improve it...working on it.

Comment (simone-silvestri) [reply src/BoundaryConditions/open_boundary_schemes/tidal_forcing.jl:90]: if it is not no worries, just seems rather expensive to me :)

**Comment by enri66** [comment]
@simone-silvestri

On `equilibrium_tide`: yes, worth it — with all ten NOAA constituents (4 semidiurnal, 4 diurnal, 2 long-period), the species' latitude structure was being recomputed once per constituent even though only three distinct values exist, and this runs twice per cell (the discrete `∂x`/`∂y` derivative) every RK stage. Fixed by hoisting the three `cos²φ`/`sin2φ`/`½−3⁄2sin²φ` evaluations out of the loop and selecting between them with `ifelse`.

On the boundary conditions in `test_tidal_body_force`: checked directly — with no `boundary_conditions` passed, this grid's defaults resolve to `NormalFlowBoundaryCondition` (impenetrable) on the normal component and free-slip on the tangential one, i.e. a fully closed basin. That's deliberate for this test: it isolates `tidal_forcing` (the body force) from `tidal_boundary_conditions`, since a closed basin this small should just fill to the equilibrium tide statically, which is exactly what's checked. But that does mean `tidal_boundary_conditions` has no dedicated test anywhere in this PR — only the body-force half is exercised. Want a second test added for that, or is coverage elsewhere (e.g. downstream in NumericalEarth) enough for now?

**Comment by simone-silvestri** [comment]
I think coverage is fine, if you resolve the conflicts I can merge

**Comment by enri66** [comment]
@simone-silvestri Two CI failures here, neither traceable to this PR's diff:

- `cpu-all-tests` failed on an `ExplicitImports` quality-assurance check in `OceananigansZarrExt` (unrelated files — this PR only touches tidal forcing/BCs). #5964, built from essentially the same merged-`main` tree around the same time, passed the identical check cleanly, so this looks like a flaky CI run rather than a real regression.
- `oceananigans-dot-jl` (AMDGPU) died with `signal: terminated` after 1h22m — a timeout, not a test failure.

Since forks can't retrigger Buildkite builds ourselves — could you kick off a rebuild when you get a chance?

**Comment by giordano** [comment]
The ExplicitImports quality-assurance check in OceananigansZarrExt is fixed by #6045

**Comment by enri66** [comment]
> The ExplicitImports quality-assurance check in OceananigansZarrExt is fixed by #6045

@giordano Thanks, so I'll wait until this merges to rebuild since it's not passing tests.

**Review by simone-silvestri — APPROVED** [review]

File: test/hydrostatic_free_surface/open_boundary_conditions.jl [hunk 7]
```diff
@@ -745,6 +807,10 @@ end
         @test test_oblique_radiation_mirror_symmetry()
     end
 
+    @testset "Equilibrium tidal body force" begin
+        @test test_tidal_body_force()
+    end
+
```
Comment (giordano) [inline test/hydrostatic_free_surface/open_boundary_conditions.jl]:
```suggestion

```

File: src/BoundaryConditions/open_boundary_schemes/tidal_forcing.jl (hunk 5 above)
Comment (enri66) [reply src/BoundaryConditions/open_boundary_schemes/tidal_forcing.jl:90]: Addressed in 653d862de: the latitude structure is now computed once per point instead of once per constituent, so the per-constituent cost is a single `cos`. It's still evaluated every step, but it only depends on `t` through that one term.
